modules/imgcodecs/src/grfmt_tiff.cpp

```diff
@@ -290,7 +290,7 @@ bool TiffDecoder::nextPage()
 
 bool  TiffDecoder::readData( Mat& img )
 {
-    if(m_hdr && img.type() == CV_32FC3)
+    if(m_hdr && img.type() == CV_32F)
```

Comment:
This breaks existed behavior (but OpenCV has a lack of tests to detect this - original patch doesn't have them too).
1. This should be fixed.
2. Test for new functionality should be added (put some small images (8x8 is enough) in your format into opencv_extra repository somewhere [here](https://github.com/opencv/opencv_extra/tree/master/testdata/highgui/readwrite)).

Regarding of this code block it should be:
```
if(m_hdr && img.type() == CV_32FC3)
{
    return readData_32FC3(img); // rename old method ("Hdr" suffix is wrong anyway)
}
if(m_hdr && img.type() == CV_32FC1)
{
    return readData_32FC1(img); // add new method with your code
}
```

BTW, Perhaps you need to fix m_type in `readHeader()` too.